// rosaenlg/dist/ChoosebestManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const synonym_optimizer_1 = require("synonym-optimizer");
class ChoosebestManager {
    constructor(language, saveRollbackManager, randomManager, defaultAmong) {
        this.language = language;
        this.saveRollbackManager = saveRollbackManager;
        this.randomManager = randomManager;
        this.defaultAmong = defaultAmong;
    }
    setSpy(spy) {
        this.spy = spy;
    }
    runChoosebest(which, params) {
        if (this.spy.isEvaluatingChoosebest()) {
            const err = new Error();
            err.name = 'InvalidArgumentError';
            err.message = `choosebest cannot be imbricated`;
            throw err;
        }
        const debugOn = params && params.debug ? true : false;
        if (debugOn) {
            params.debugRes = {
                maxTest: null,
                perfectScoreAfter: null,
                bestScore: null,
                bestText: null,
                bestDebug: null,
                worstScore: null,
                worstText: null,
                worstDebug: null,
            };
        }
        let maxTest;
        if (params && params.among) {
            maxTest = params.among;
        }
        else {
            maxTest = this.defaultAmong;
        }
        if (debugOn) {
            params.debugRes.maxTest = maxTest;
        }
        function getFromParamsAndStoreDebug(paramName) {
            const res = params && params[paramName] ? params[paramName] : null;
            if (debugOn && res) {
                params.debugRes[paramName] = res;
            }
            return res;
        }
        const stopWordsAdd = getFromParamsAndStoreDebug('stop_words_add');
        const stopWordsRemove = getFromParamsAndStoreDebug('stop_words_remove');
        const stopWordsOverride = getFromParamsAndStoreDebug('stop_words_override');
        const identicals = getFromParamsAndStoreDebug('identicals');
        const newContentStart = this.spy.getPugHtml().length;
        // SIMULATE
        const scores = [];
        let alternatives;
        let debugInfos;
        if (debugOn) {
            alternatives = [];
            debugInfos = [];
        }
        for (let i = 0; i < maxTest; i++) {
            // SAVE
            this.saveRollbackManager.saveSituation('choosebest');
            this.randomManager.incrRnd(i);
            this.spy.getPugMixins()[which](params);
            const generated = this.spy.getPugHtml().substring(newContentStart);
            // ROLLBACK
            this.saveRollbackManager.rollback();
            if (debugOn) {
                alternatives.push(generated);
            }
            const debugInfo = {
                filteredAlt: null,
                identicals: null,
                identicalsMap: null,
                wordsWithPos: null,
                score: null,
            };
            const score = synonym_optimizer_1.scoreAlternative(this.language, generated, stopWordsAdd, stopWordsRemove, stopWordsOverride, identicals, debugOn ? debugInfo : null);
            scores.push(score);
            if (debugOn) {
                debugInfos.push(debugInfo);
            }
            // console.log(`${generated} => ${score}`);
            // we can stop before if we ever get a perfect score
            if (score === 0) {
                if (debugOn) {
                    params.debugRes.perfectScoreAfter = i;
                }
                break;
            }
        }
        // CHOOSE BEST
        const best = scores.indexOf(Math.min(...scores));
        if (debugOn) {
            params.debugRes.bestScore = scores[best];
            params.debugRes.bestText = alternatives[best];
            params.debugRes.bestDebug = debugInfos[best];
        }
        if (debugOn) {
            const worst = scores.indexOf(Math.max(...scores));
            params.debugRes.worstScore = scores[worst];
            params.debugRes.worstText = alternatives[worst];
            params.debugRes.worstDebug = debugInfos[worst];
        }
        // CHANGE RANDOM POSITION
        this.randomManager.incrRnd(best);
        // AND GENERATE IT
        this.spy.getPugMixins()[which](params);
    }
}
exports.ChoosebestManager = ChoosebestManager;
