// lefff-helper/dist/index.js
"use strict";
//import * as Debug from 'debug';
//const debug = Debug("lefff-helper");
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
class LefffHelper {
    isAdj(ff) {
        return this.getAdj(ff) != null;
    }
    isNoun(ff) {
        return this.getNoun(ff) != null;
    }
    getNoun(ff) {
        // debug(`looking for noun ${ff}`);
        if (!this.nouns) {
            this.nouns = JSON.parse(fs_1.readFileSync(__dirname + '/../resources_pub/nouns.json', 'utf8'));
        }
        return this.nouns[ff];
    }
    getAdj(ff) {
        if (!this.adjectives) {
            this.adjectives = JSON.parse(fs_1.readFileSync(__dirname + '/../resources_pub/adjectives.json', 'utf8'));
        }
        const adjectiveInfo = this.adjectives[ff];
        if (!adjectiveInfo) {
            return null;
        }
        const racine = adjectiveInfo[0];
        const isPp = adjectiveInfo[1];
        if (isPp) {
            /*
              c'est un participe passé
              on ne veut pas le verbe mais la forme ms de ce même participe passé
            */
            if (!this.pastParticiples) {
                this.pastParticiples = JSON.parse(fs_1.readFileSync(__dirname + '/../resources_pub/pastParticiples.json', 'utf8'));
            }
            return this.pastParticiples[racine];
        }
        else {
            return racine;
        }
    }
}
exports.LefffHelper = LefffHelper;
