// rosaenlg-packager/dist/rosaenlg-packager-bundle.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('fs'), require('path')) :
	typeof define === 'function' && define.amd ? define(['exports', 'fs', 'path'], factory) :
	(global = global || self, factory(global.rosaenlgPackager = {}, global.fs, global.path));
}(this, (function (exports, fs, path) { 'use strict';

	fs = fs && Object.prototype.hasOwnProperty.call(fs, 'default') ? fs['default'] : fs;
	path = path && Object.prototype.hasOwnProperty.call(path, 'default') ? path['default'] : path;

	var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

	function commonjsRequire () {
		throw new Error('Dynamic requires are not currently supported by @rollup/plugin-commonjs');
	}

	function unwrapExports (x) {
		return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
	}

	function createCommonjsModule(fn, module) {
		return module = { exports: {} }, fn(module, module.exports), module.exports;
	}

	function getCjsExportFromNamespace (n) {
		return n && n['default'] || n;
	}

	var jsPackager = createCommonjsModule(function (module, exports) {
	"use strict";
	Object.defineProperty(exports, "__esModule", { value: true });
	function compileTemplateToJsString(entryTemplate, language, staticFs, rosaeNlgFeatures, exportDefault = false, compileDebug = false) {
	    const options = {
	        embedResources: true,
	        language: language,
	        compileDebug: compileDebug,
	    };
	    if (staticFs != null) {
	        options.staticFs = staticFs;
	    }
	    const compiled = rosaeNlgFeatures.compileFileClient(entryTemplate, options);
	    const res = exportDefault ? compiled + `\nexport default template;` : compiled;
	    return res;
	}
	exports.compileTemplateToJsString = compileTemplateToJsString;

	});

	var jsPackager$1 = unwrapExports(jsPackager);
	var jsPackager_1 = jsPackager.compileTemplateToJsString;

	var jsonPackager = createCommonjsModule(function (module, exports) {
	"use strict";
	Object.defineProperty(exports, "__esModule", { value: true });


	const FORMAT = '2.0.0';
	function getFinalFileName(baseDir, template) {
	    const pathElts = path.parse(template);
	    const templateWithExt = pathElts.ext == null || pathElts.ext == '' ? template + '.pug' : template;
	    const fullPath = baseDir ? path.join(baseDir, templateWithExt) : templateWithExt;
	    const finalFileName = fullPath.replace(new RegExp('\\\\', 'g'), '/'); // change to linux paths
	    return finalFileName;
	}
	const includeRe = new RegExp('^include\\s+(.+)$');
	function getTemplatesMap(baseDir, template, templatesMap) {
	    //console.log('starting to process: ' + template);
	    templatesMap = templatesMap || {};
	    const finalFileName = getFinalFileName(baseDir, template);
	    // console.log('finalFileName is: ' + finalFileName);
	    const content = fs.readFileSync(finalFileName, 'utf8');
	    // add this one
	    templatesMap[finalFileName] = content;
	    // check includes
	    const lines = content.split(/\r?\n/);
	    for (let i = 0; i < lines.length; i++) {
	        const line = lines[i];
	        const matches = line.trim().match(includeRe);
	        if (matches && matches[1]) {
	            const matched = matches[1];
	            // console.log('found included: ' + matched);
	            const newBaseDir = path.parse(finalFileName).dir;
	            // console.log('newBaseDir: ' + newBaseDir);
	            templatesMap = getTemplatesMap(newBaseDir, matched, templatesMap);
	        }
	    }
	    return templatesMap;
	}
	function completePackagedTemplateJson(packagedTemplate, rosaeNlgFeatures) {
	    packagedTemplate.format = FORMAT;
	    // if templates are not here, we read them from the disk
	    if (!packagedTemplate.src.templates) {
	        packagedTemplate.src.templates = getTemplatesMap(null, packagedTemplate.src.entryTemplate, null);
	    }
	    // compile if asked
	    if (packagedTemplate.src.compileInfo && packagedTemplate.src.compileInfo.activate) {
	        const options = {
	            embedResources: true,
	            language: packagedTemplate.src.compileInfo.language,
	            compileDebug: packagedTemplate.src.compileInfo.compileDebug,
	            staticFs: packagedTemplate.src.templates,
	        };
	        const compiled = rosaeNlgFeatures.compileFileClient(packagedTemplate.src.entryTemplate, options);
	        packagedTemplate.comp = {
	            compiled: compiled,
	            compiledWithVersion: rosaeNlgFeatures.getRosaeNlgVersion(),
	            compiledBy: 'rosaenlg-packager',
	            compiledWhen: new Date().toISOString(),
	        };
	    }
	}
	exports.completePackagedTemplateJson = completePackagedTemplateJson;

	});

	var jsonPackager$1 = unwrapExports(jsonPackager);
	var jsonPackager_1 = jsonPackager.completePackagedTemplateJson;

	var dist = createCommonjsModule(function (module, exports) {
	"use strict";
	Object.defineProperty(exports, "__esModule", { value: true });

	exports.compileTemplateToJsString = jsPackager.compileTemplateToJsString;

	exports.completePackagedTemplateJson = jsonPackager.completePackagedTemplateJson;

	});

	var index = unwrapExports(dist);
	var dist_1 = dist.compileTemplateToJsString;
	var dist_2 = dist.completePackagedTemplateJson;

	exports.compileTemplateToJsString = dist_1;
	exports.completePackagedTemplateJson = dist_2;
	exports.default = index;

	Object.defineProperty(exports, '__esModule', { value: true });

})));
